cuqi/geometry/_geometry.py

```diff
@@ -672,15 +685,49 @@ def normalizer(self):
 
     @property
     def coefs(self):
+        # Return None if self.num_modes is 0,
+        # that is when num_modes is not provided and
+        # the grid is None (fun_dim is None).
+        if self.num_modes == 0:
+            return None
+
+        # If the coefficients are not computed, compute them.
+        if self._coefs is None or len(self._coefs) != self.num_modes:
+            eigvals = np.array(range(1, self.par_dim+1))  # KL eigvals
+            self._coefs = 1/np.float_power(eigvals, self.decay_rate)
+
+        # Return the coefficients.
         return self._coefs
 
+    @property
+    def num_modes(self):
+        # If grid is not set, interpret grid dimension as 0.
```

Comment:
Same for this one, please first explain what the method does in general before the special case.

Reply:
here too